frontend/test/resolution/testArrays.cpp

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2021-2023 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "test-resolution.h"
+#include "test-minimal-modules.h"
+
+#include "chpl/parsing/parsing-queries.h"
+#include "chpl/resolution/resolution-queries.h"
+#include "chpl/resolution/scope-queries.h"
+#include "chpl/types/all-types.h"
+#include "chpl/uast/Identifier.h"
+#include "chpl/uast/Module.h"
+#include "chpl/uast/Record.h"
+#include "chpl/uast/Variable.h"
+
+// TODO:
+// - this[] access
+// - Slices
+
+static void testArray(std::string domainType,
+                      std::string eltType) {
+  Context ctx;
+  Context* context = &ctx;
+  ErrorGuard guard(context);
+
+  // a different element type from the one we were given
+  std::string altElt = "int";
+  if (eltType == "int") {
+    altElt = "string";
+  }
+
+  std::string program = DomainModule + ArrayModule +
+R"""(
+module M {
+  use ChapelArray;
+  
+  var d : )""" + domainType + R"""(;
+  type eltType = )""" + eltType + R"""(;
+
+  var A : [d] eltType;
+
+  const AD = A.domain;
+  const s = A.size;
+
+  for loopI in A {
+    var z = loopI;
+  }
+
+  proc generic(arg: []) {
+    type GT = arg.type;
+    return 42;
+  }
+
+  proc eltOnly(arg: [] )""" + eltType + R"""() param {
+    type ETGood = arg.type;
+    return "correct method";
+  }
+
+  proc eltOnly(arg: [] )""" + altElt + R"""() param {
+    type ETBad = arg.type;
+    return "wrong method";
+  }
+
+  var g_ret = generic(A);
+  param e_ret = eltOnly(A);
+}
+)""";
+
+  auto path = UniqueString::get(context, "input.chpl");
+  setFileText(context, path, std::move(program));
+
+  const ModuleVec& vec = parseToplevel(context, path);
+  const Module* m = vec[2]; 
+
+  const ResolutionResultByPostorderID& rr = resolveModule(context, m->id());
+
+  const Variable* d = findOnlyNamed(m, "d")->toVariable();
+  QualifiedType dType = rr.byAst(d).type();
```

Comment:
This "findOnlyNamed -> toVariable -> byAst -> Type" thing could be made a helper function to clean up some code